src/Orleans/Utils/Pool/ConcurrentObjectPool.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Orleans.Runtime
+{
+    public class ConcurrentObjectPool<T> : IObjectPool<T>
+        where T : PooledResource<T>, IDisposable
+    {
+        private readonly ConcurrentBag<T> _objects;
+        private readonly Func<T> _objectGenerator;
+
+        public ConcurrentObjectPool(Func<T> objectGenerator)
```

Comment:
Suggest adding an initial size and pre-allocating.  This should reduce warmup cost and allow GC to compact messages into gen2 more easily.